File: src/main/java/net/vpg/game2048/Board.java
package net.vpg.game2048;

import java.util.Arrays;
import java.util.Optional;
import java.util.Random;
import java.util.stream.Stream;

public class Board {
    final int size;
    final Random random;
    Cell[][] cells;
    int score;

    public Board(int size) {
        this.size = size;
        this.cells = new Cell[size][size];
        for (int i = 0; i < size; i++) {
            for (int j = 0; j < size; j++) {
                new Cell(i, j, this);
            }
        }
        this.random = new Random();
    }

    public Cell[][] getCells() {
        return cells;
    }

    public Stream<Cell> getCellsAsStream() {
        return Arrays.stream(cells).flatMap(Arrays::stream);
    }

    public int move(Move move) {
        int row = move.row * (size - 1);
        int column = move.column * (size - 1);
        for (int runs = 0; runs < size; runs++) {
            cells[row][column].move(move);
            row += move.rowChange;
            column += move.columnChange;
        }
        spawn();
        int moveScore = getCellsAsStream()
            .filter(Cell::isModified)
            .peek(Cell::removeModifyFlag)
            .mapToInt(Cell::getValue)
            .sum();
        score += moveScore;
        return moveScore;
    }

    public boolean checkLose() {
        return getCellsAsStream().noneMatch(Cell::canMove);
    }

    public boolean checkWin() {
        return getCellsAsStream().anyMatch(Cell::isFinal);
    }

    void spawn() {
        Optional.of(getCellsAsStream().filter(Cell::isEmpty).toArray(Cell[]::new))
            .filter(emptyCells -> emptyCells.length != 0)
            .ifPresent(emptyCells -> Spawner.getInstance().spawn(emptyCells[random.nextInt(emptyCells.length)]));
    }

    public int getScore() {
        return score;
    }

    public String toString() {
        StringBuilder tor = new StringBuilder();
        tor.append('+');
        tor.append("------+".repeat(size));
        int firstLineLen = tor.length();
        tor.append('\n');
        for (Cell[] row : cells) {
            tor.append('|');
            for (Cell cell : row) {
                tor.append(' ').append(cell.getFormatted()).append(' ').append('|');
            }
            tor.append('\n');
        }
        tor.append(tor, 0, firstLineLen);
        return tor.toString();
    }
}

File: src/main/java/net/vpg/game2048/Board.kt
package net.vpg.game2048

import net.vpg.game2048.Spawner.spawn
import kotlin.math.absoluteValue

class Board(val size: Int) {
    val cells = Array<Array<Cell>>(size) { i ->
        Array<Cell>(size) { j ->
            Cell(i, j, this)
        }
    }
    val flatCells = cells.flatten()
    var score = 0
    var win = false
    var lose = false

    fun move(move: Move): Int {
        var row = move.row * (size - 1)
        var column = move.column * (size - 1)
        repeat(size, {
            cells[row][column].move(move)
            row += move.columnChange.absoluteValue
            column += move.rowChange.absoluteValue
        })
        val moveScore = flatCells
            .filter { it.isModified }
            .onEach { it.isModified = false }
            .sumOf { it.value }
        val anyMove = flatCells
            .filter { it.isMoved }
            .onEach { it.isMoved = false }
            .any()
        if (anyMove) {
            spawn()
        }
        win = flatCells.any { it.isFinal }
        lose = flatCells.none { it.canMove() }
        score += moveScore
        return moveScore
    }

    fun spawn() = flatCells.filter { it.isEmpty }
        .takeIf { it.isNotEmpty() }
        ?.random()
        ?.spawn()
}
